runtime/src/main/cpp/Memory.cpp

```diff
@@ -2056,51 +2056,81 @@ OBJ_GETTER(initSharedInstance,
 #endif  // KONAN_NO_THREADS
 }
 
+/**
+ * We keep thread affinity and reference value based cookie in the atomic references, so that
+ * repeating read operation of the same value do not lead to the repeating rememberNewContainer() operation.
+ * We must invalidate cookie after the local GC, as otherwise fact that container of the `value` is retained
+ * may change, if the last reference to the value read is lost during GC and we re-read same value from
+ * the same atomic reference. Thus we also include last GC timestamp into the cookie.
+ */
+inline int32_t computeCookie(ObjHeader* value) {
+  auto* state = memoryState;
+  auto timestamp = state->lastGcTimestamp;
+  return (static_cast<int32_t>(reinterpret_cast<intptr_t>(state))) ^
+    (static_cast<int32_t>(reinterpret_cast<intptr_t>(value))) ^
```

Comment:
Is it actually required to include `value` to cookie?

Reply:
That's indeed may not be needed. I was thinking about scenario where we read value, write new value and then read the new value from the same thread (and second read doesn't remember the new container), but indeed such a container must be already remembered as we managed to write it.